File: src/converter/plugins/CheckCoveragePlugin.ts
import {ConverterPlugin} from "../plugin.js";
import ts, {Node} from "typescript";
import {Render} from "../render.js";
import {ConverterContext} from "../context.js";
import {traverse} from "../../utils/traverse.js";
import {ConfigurationService, configurationServiceKey} from "./ConfigurationPlugin.js";
import {TypeScriptService, typeScriptServiceKey} from "./TypeScriptPlugin.js";

export const checkCoverageServiceKey = Symbol()

export class CheckCoverageService {
    private coveredNodes = new Set<Node>()

    private allNodes = new Set<Node>()

    register(node: Node) {
        this.allNodes.add(node)
    }

    cover(node: Node) {
        this.coveredNodes.add(node)
    }

    deepCover(node: Node) {
        traverse(node, it => this.cover(it))
    }

    emit(callback: (uncoveredNode: Node) => void): { coveredNodes: number, uncoveredNodes: number } {
        for (const node of this.allNodes){
            if (!this.coveredNodes.has(node)) {
                callback(node);
            }
        }

        const result = {
            coveredNodes: this.coveredNodes.size,
            uncoveredNodes: this.allNodes.size - this.coveredNodes.size,
        }

        this.coveredNodes = new Set<Node>()
        this.allNodes = new Set<Node>()

        return result
    }
}

export class CheckCoveragePlugin implements ConverterPlugin {
    private readonly checkCoverageService = new CheckCoverageService()

    generate(context: ConverterContext): Record<string, string> {
        const configurationService = context.lookupService<ConfigurationService>(configurationServiceKey)
        const typeScriptService = context.lookupService<TypeScriptService>(typeScriptServiceKey)

        const {coveredNodes, uncoveredNodes} = this.checkCoverageService.emit( uncoveredNode => {
            if (configurationService?.configuration?.verbose) {
                const message = `Node with kind ${ts.SyntaxKind[uncoveredNode.kind]} is uncovered`
                const sourceFile = uncoveredNode.getSourceFile()

                if (sourceFile) {
                    const {line, character} = sourceFile.getLineAndCharacterOfPosition(uncoveredNode.pos)

                    console.error(`${sourceFile.fileName}: (${line + 1}, ${character + 1}): ${message}`)
                } else {
                    console.error(message)
                }

                console.error("--- Node Start ---");
                console.error(typeScriptService?.printNode(uncoveredNode));
                console.error("--- Node End ---");

                console.error();
            }
        })

        console.log(`Covered nodes: ${coveredNodes}`)
        console.log(`Uncovered nodes: ${uncoveredNodes}`)

        return {};
    }

    render(node: Node, context: ConverterContext, next: Render): string | null {
        return null;
    }

    traverse(node: Node): void {
        this.checkCoverageService.register(node)
    }

    setup(context: ConverterContext): void {
        context.registerService(checkCoverageServiceKey, this.checkCoverageService)
    }
}

File: gradle-plugin/karakum/src/jsMain/kotlin/io/github/sgrishchenko/karakum/extension/plugins/CheckCoveragePlugin.kt
package io.github.sgrishchenko.karakum.extension.plugins

import io.github.sgrishchenko.karakum.extension.ConverterContext
import io.github.sgrishchenko.karakum.extension.ConverterPlugin
import io.github.sgrishchenko.karakum.extension.GeneratedFile
import io.github.sgrishchenko.karakum.extension.Render
import io.github.sgrishchenko.karakum.util.getSourceFileOrNull
import io.github.sgrishchenko.karakum.util.traverse
import js.array.ReadonlyArray
import js.objects.JsPlainObject
import js.objects.ReadonlyRecord
import js.symbol.Symbol
import typescript.Node
import typescript.SyntaxKind

val checkCoverageServiceKey = Symbol()

@JsPlainObject
external interface CheckCoverageResult {
    val coveredNodes: Int
    val uncoveredNodes: Int
}

class CheckCoverageService {
    private val coveredNodes = mutableSetOf<Node>()

    private val allNodes = mutableSetOf<Node>()

    fun register(node: Node) {
        this.allNodes.add(node)
    }

    fun cover(node: Node) {
        this.coveredNodes.add(node)
    }

    fun deepCover(node: Node) {
        traverse(node) { cover(it) }
    }

    fun emit(callback: (uncoveredNode: Node) -> Unit): CheckCoverageResult {
        for (node in allNodes){
            if (node !in coveredNodes) {
                callback(node);
            }
        }

        val result = CheckCoverageResult(
            coveredNodes = coveredNodes.size,
            uncoveredNodes = allNodes.size - coveredNodes.size,
        )

        coveredNodes.clear()
        allNodes.clear()

        return result
    }
}

class CheckCoveragePlugin : ConverterPlugin<Node> {
    private val checkCoverageService = CheckCoverageService()

    override fun generate(context: ConverterContext, render: Render<Node>): ReadonlyArray<GeneratedFile> {
        val configurationService = context.lookupService<ConfigurationService>(configurationServiceKey)
        val typeScriptService = context.lookupService<TypeScriptService>(typeScriptServiceKey)

        val result = this.checkCoverageService.emit{ uncoveredNode ->
            if (configurationService?.configuration?.verbose == true) {
                val syntaxKindRecord = SyntaxKind.unsafeCast<ReadonlyRecord<SyntaxKind, String>>()
                val message = "Node with kind ${syntaxKindRecord[uncoveredNode.kind]} is uncovered"
                val sourceFile = uncoveredNode.getSourceFileOrNull()

                if (sourceFile != null) {
                    val lineAndCharacter = sourceFile.getLineAndCharacterOfPosition(uncoveredNode.pos)
                    val line = lineAndCharacter.line
                    val character = lineAndCharacter.character

                    console.error("${sourceFile.fileName}: (${line + 1}, ${character + 1}): $message")
                } else {
                    console.error(message)
                }

                console.error("--- Node Start ---");
                console.error(typeScriptService?.printNode(uncoveredNode));
                console.error("--- Node End ---");

                console.error();
            }
        }

        console.log("Covered nodes: ${result.coveredNodes}")
        console.log("Uncovered nodes: ${result.uncoveredNodes}")

        return emptyArray()
    }

    override fun render(node: Node, context: ConverterContext, next: Render<Node>) = null

    override fun traverse(node: Node, context: ConverterContext) {
        this.checkCoverageService.register(node)
    }

    override fun setup(context: ConverterContext) {
        context.registerService(checkCoverageServiceKey, this.checkCoverageService)
    }
}
